### DISCUSSION TITLE
Status of the Go language bindings

### DISCUSSION THREAD

**Discussion by benozol** [bytecodealliance/wasm-micro-runtime#2672] [General]
`wasm-micro-runtime` provides [bindings for the Go language](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/language-bindings/go), but it seems that they are currently not in focus of the development:

- the Go module name in `go.mod` doesn't match the organization, which makes `go get` choke on it,
- parts of the C API are unavailable (e.g. the interpreter modes),
- there weren't any updates since January, and
- I was unable to run the example in the [README](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/language-bindings/go/README.md) (missing result value after calling the Wasm function).

Because we are currently looking for a Wasm interpreter (not JIT) in Go, I would like to inquire about the current plans for these bindings. Are there updates planned? Which target platforms will be supported? Will the `wamr` interpreter mode become available? Or does is just depend on external contributions?

**Comment by lum1n0us** [comment]
*interpreter* works fine. *go binding* is lack of maintenance for a while. Please feel free to contribute.

**Comment by wenyongh** [comment]
Hi @benozol thanks for paying attention to WAMR's go language binding:

> `wasm-micro-runtime` provides [bindings for the Go language](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/language-bindings/go), but it seems that they are currently not in focus of the development:
> 
> * the Go module name in `go.mod` doesn't match the organization, which makes `go get` choke on it,

We are not familiar with how to publish the the go package, per my understanding, we should put the header and binary files under a specify folder, for example:
```
  <wamr_root>/language-bindings/go/packed/include: header files
  <wamr_root>/language-bindings/go/packed/lib/linux-amd64: dummy.go  libvmlib.a
```
and in go.mod, add "module github.com/bytecodealliance/wasm-micro-runtime/language-bindings/go", right?

> * parts of the C API are unavailable (e.g. the interpreter modes),

Yes, but many C APIs have been supported, it can meet a lot of requirements. You can add more if needed, and open issue in community for help if needed.

> * there weren't any updates since January, and

It wasn't updated, but the samples work fine.

> * I was unable to run the example in the [README](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/language-bindings/go/README.md) (missing result value after calling the Wasm function).

Please cd samples and run `./builds.h`, it will build the sample and run it, below is some of the output:
```bash
Call func `i32 fib2(i32)` with CallFunc:
fib2(32) return: 3524578

Call func `void test1(i32, i64, f32, f64)` with CallFuncV:
i32: 12345678, i64: 3344556677889900, f32: 5678.123535, f64: 987654321.567800

Call func `i64 test2(i64, i64)` with CallFuncV:
3344556677889900 + 1122331122110099 = 4466887799999999
test2(3344556677889900, 1122331122110099) return: 4466887799999999
```
Note that **the result is gotten and dumped after calling some functions**.

**Comment by benozol** [comment]
Thanks for your reply, that's reassuring! I will try the bindings in our use case.

> and in go.mod, add "module github.com/bytecodealliance/wasm-micro-runtime/language-bindings/go", right?

Yes, that's my understanding

> Note that the result is gotten and dumped after calling some functions.

Great, I will check what went wrong on my side

**Comment by benozol** [comment]
Figured out why my test wasn't working: `i32` results were not returned due to Go's switch semantics:

- https://github.com/bytecodealliance/wasm-micro-runtime/pull/2676
